model/generate.go

```diff
@@ -102,65 +102,92 @@ func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
 
 // NewVersion adds the buildvariants, tasks, and functions
 // from a generated project config to a project, and returns the previous config string.
-func (g *GeneratedProject) NewVersion() (*Project, *Version, *task.Task, *projectMaps, string, error) {
+func (g *GeneratedProject) NewVersion() (*Project, *Version, *task.Task, *projectMaps, *parserProject, error) {
 	// Get task, version, and project.
 	t, err := task.FindOneId(g.TaskID)
 	if err != nil {
-		return nil, nil, nil, nil, "",
+		return nil, nil, nil, nil, nil,
 			gimlet.ErrorResponse{StatusCode: http.StatusInternalServerError, Message: errors.Wrapf(err, "error finding task %s", g.TaskID).Error()}
 	}
 	if t == nil {
-		return nil, nil, nil, nil, "",
+		return nil, nil, nil, nil, nil,
 			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrapf(err, "unable to find task %s", g.TaskID).Error()}
 	}
 	v, err := VersionFindOneId(t.Version)
 	if err != nil {
-		return nil, nil, nil, nil, "",
+		return nil, nil, nil, nil, nil,
 			gimlet.ErrorResponse{StatusCode: http.StatusInternalServerError, Message: errors.Wrapf(err, "error finding version %s", t.Version).Error()}
 	}
 	if v == nil {
-		return nil, nil, nil, nil, "",
+		return nil, nil, nil, nil, nil,
 			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Wrapf(err, "unable to find version %s", t.Version).Error()}
 	}
-	if v.Config == "" {
-		return nil, nil, nil, nil, "",
-			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Errorf("unable to find config string for version %s", t.Version).Error()}
+	if v.Config == "" && v.Project == nil {
+		return nil, nil, nil, nil, nil,
+			gimlet.ErrorResponse{StatusCode: http.StatusBadRequest, Message: errors.Errorf("neither a config string nor a project is present for version %s", t.Version).Error()}
+	}
+	var p *Project
+	var prevProject *parserProject    // nil if v.Project is unset, used to ensure updates are atomic
+	var currentProject *parserProject // not nil, because will be set from config string if v.Project is unset
+	var errs []error
+	if v.Project == nil {
+		if currentProject, errs = CreateIntermediateProject([]byte(v.Config)); len(errs) != 0 {
```

Comment:
CreateIntermediateProject should return a single error, processed with a grip.Catcher rather than this slice of errors.